pkg/generate/output/artifact.go

```diff
@@ -0,0 +1,178 @@
+	// The checksum of all the combined files generated within the APIs directory
+	ApisDirectoryChecksum string `json:"apis_directory_checksum"`
+	// Last modification reason
+	LastModificationReason UpdateReason `json:"last_modification_reason"`
+	// Last modification time
+	LastModificationTime string `json:"last_modification_time"`
+
+	// AWS SDK Go version used generate the APIs
+	AWSSDKGoVersion string `json:"aws_sdk_go_version"`
+
+	// Informatiom about the ack-generate binary used to generate the APIs
+	ACKGenerateInfo ackGenerateInfo `json:"ack_generate_info"`
+	// Information about the generator config file used to generate the APIs
+	GeneratorConfigInfo generatorConfigInfo `json:"generator_config_info"`
+}
+
+// ack-generate binary information
+type ackGenerateInfo struct {
+	Version   string `json:"version"`
+	GoVersion string `json:"go_version"`
+	BuildDate string `json:"build_date"`
+	BuildHash string `json:"build_hash"`
+}
+
+// generator.yaml information
+type generatorConfigInfo struct {
+	OriginalFileName string `json:"original_file_name"`
+	FileChecksum     string `json:"file_checksum"`
+}
+
+// CreateArtifact gathers information about the generated code and save
+// a yaml version in the API version directory
+func CreateArtifact(
+	apisVersion string,
+	apisPath string,
+	modificationReason UpdateReason,
+	awsSDKGo string,
+	generatorFileName string,
+) error {
+	filesDirectory := filepath.Join(apisPath, apisVersion)
+	hash, err := hashDirectoryContent(filesDirectory)
+	if err != nil {
+		return err
+	}
+
+	generatorFileHash, err := hashFile(generatorFileName)
+	if err != nil {
+		return err
+	}
+
+	artifact := &Artifact{
+		ApisVersion:            apisVersion,
+		ApisDirectoryChecksum:  hash,
+		LastModificationTime:   time.Now().UTC().String(),
+		LastModificationReason: modificationReason,
+		AWSSDKGoVersion:        awsSDKGo,
+		ACKGenerateInfo: ackGenerateInfo{
+			Version:   version.Version,
+			BuildDate: version.BuildDate,
+			BuildHash: version.BuildHash,
+			GoVersion: version.GoVersion,
+		},
+		GeneratorConfigInfo: generatorConfigInfo{
+			OriginalFileName: filepath.Base(generatorFileName),
+			FileChecksum:     generatorFileHash,
+		},
+	}
+
+	data, err := yaml.Marshal(artifact)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(filepath.Join(filesDirectory, outputFileName), data, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// hashDirectoryContent returns the sha1 checksum of a given directory. It will walk
+// the file tree of a directory and combine and the file contents before hashing it.
+func hashDirectoryContent(directory string) (string, error) {
+	h := sha1.New()
+	err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			fileReader, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+			_, err = io.Copy(h, fileReader)
+			if err != nil {
+				return err
+			}
```

Comment:
Wow I'm surprised at how expensive this operation is :/

Reply:
Yes it is...

Reply:
I just realized that i forgot to ignore `output.yaml` and `generator.yaml` :)